File: nowcoder/question/NC50.cpp
class Solution {
public:
    /**
     *
     * @param head ListNode类
     * @param k int整型
     * @return ListNode类
     */
     //区间[start,end)
    ListNode* reverseInterval(ListNode* start, ListNode* end) {
        ListNode* pre = nullptr, * cur = start;
        while (cur != end) {
            ListNode* nxt = cur->next;
            cur->next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }


    ListNode* reverseKGroup(ListNode* head, int k) {
        if (!head) return head;

        ListNode* a = head, * b = head;
        for (int i = 0; i < k; i++) {
            if (b == nullptr) return head;
            b = b->next;
        }

        ListNode* newHead = reverseInterval(a, b);
        a->next = reverseKGroup(b, k);
        return newHead;
    }
};

File: 反转链表/nowcoder/NC 50.js
function reverseKGroup( head ,  k ) {
    // write code here
    function shouldReverse(node){
        let temp = node
        let count = 0
        if(!node){
            return false
        }
        
        while(temp){
            if(count === k){
                return true
            }
            temp = temp.next
            ++count
        }
        
        return false
    }
    
    function reverse(preNode){
        let cur = preNode.next
        for(let i = 0; i < k - 1; ++i){
            let nxt = cur.next
            cur.next = nxt.next
            nxt.next = preNode.next
            preNode.next = nxt
        }
    }
    
    if(k < 2 || !head || !head.next){
        return head
    }
    
    let dummy = new ListNode(-1)
    dummy.next = head
    let cur = dummy
    if(!shouldReverse(dummy)){
        return dummy.next
    }
    
    while(shouldReverse(cur)){
        reverse(cur)
        for(let i = 0; i < k; ++i){
            cur = cur.next
        }
    }
    
    return dummy.next
}